<a href="https://colab.research.google.com/github/harshajacob1995-dev/Mutual-Fund-Stock-Analysis/blob/main/mutual_funds_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Mutual Fund Plan With Python

## Importing libraries & Loading file

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from google.colab import files

In [2]:
f = files.upload()

Saving nifty50_closing_prices - nifty50_closing_prices.csv to nifty50_closing_prices - nifty50_closing_prices.csv


In [3]:
data = pd.read_csv('nifty50_closing_prices - nifty50_closing_prices.csv')

In [5]:
data

Date  ADANIENT.NS  ADANIPORTS.NS  APOLLOHOSP.NS  ASIANPAINT.NS  \
0    2024-01-01  2914.533936    1038.321777    5727.949707    3327.104004   
1    2024-01-02  2929.719971    1068.593994    5724.263184    3322.450439   
2    2024-01-03  3000.255371    1084.299805    5746.128906    3305.061035   
3    2024-01-04  2995.559814    1112.986572    5740.052734    3311.281982   
4    2024-01-05  3003.852051    1143.754272    5732.133301    3287.377686   
..          ...          ...            ...            ...            ...   
343  2025-05-26  2546.397949    1393.420654    7096.500000    2305.925049   
344  2025-05-27  2539.001709    1397.401367    7073.500000    2306.321533   
345  2025-05-28  2513.714600    1404.168457    6956.500000    2282.144531   
346  2025-05-29  2539.501465    1431.634766    6923.000000    2273.325928   
347  2025-05-30  2518.612061    1425.862915    6880.500000    2238.447754   

     AXISBANK.NS  BAJAJ-AUTO.NS  BAJAJFINSV.NS  BAJFINANCE.NS  BHARTIARTL.NS  \
0    1096.853394    6647.077148    1675.343018     659.398193    1007.435730   
1    1092.406982    6611.317871    1685.136841     671.232849    1016.037842   
2    1097.752808    6930.176270    1676.592285     667.144897    1029.463013   
3    1122.533691    6866.345215    1700.876831     696.121582    1042.092651   
4    1136.073120    6922.240723    1708.422119     696.627441    1040.799805   
..           ...            ...            ...            ...            ...   
343  1215.199951    8958.500000    2052.600098     882.636902    1847.400024   
344  1195.599976    8963.000000    2029.900024     873.257752    1843.800049   
345  1194.699951    8849.000000    2022.400024     882.446411    1856.099976   
346  1201.699951    8874.500000    2010.900024     876.400024    1863.599976   
347  1192.199951    8607.000000    2017.400024     918.049988    1856.199951   

     ...  SHREECEM.NS  SUNPHARMA.NS  TATAMOTORS.NS  TATASTEEL.NS       TCS.NS  \
0    ...  28304.26367   1241.662476     781.448975    133.963882  3665.432861   
1    ...  27933.27734   1277.082153     775.320801    133.724396  3638.599121   
2    ...  28065.74023   1279.446777     772.404968    129.653290  3550.644531   
3    ...  28201.13867   1293.585205     786.539429    128.599579  3526.648193   
4    ...  26873.91602   1281.023193     781.794983    128.024826  3595.030762   
..   ...          ...           ...            ...           ...          ...   
343  ...  31100.00000   1676.099976     722.785584    158.806534  3506.136230   
344  ...  30950.00000   1683.400024     710.342590    157.975922  3467.876221   
345  ...  29895.00000   1666.400024     711.829834    157.565491  3466.885010   
346  ...  29915.00000   1699.800049     718.274414    159.402634  3467.281738   
347  ...  29595.00000   1677.599976     713.366577    157.350494  3432.887207   

        TECHM.NS     TITAN.NS  ULTRACEMCO.NS      UPL.NS    WIPRO.NS  
0    1262.217896  3665.156982   10402.311520  592.700012  233.428848  
1    1243.680786  3686.188232   10144.540040  594.450012  229.539612  
2    1210.693848  3681.005371   10051.939450  586.849976  223.033035  
3    1213.904907  3707.568359    9955.759766  587.650024  221.320785  
4    1214.683472  3700.441650    9957.001953  581.250000  223.375488  
..           ...          ...            ...         ...         ...  
343  1601.400024  3610.300049   11689.000000  632.049988  250.240005  
344  1592.400024  3589.399902   11421.000000  630.349976  248.250000  
345  1580.300049  3574.699951   11238.000000  630.750000  247.429993  
346  1600.000000  3588.300049   11269.000000  631.000000  250.139999  
347  1573.900024  3555.000000   11210.000000  627.900024  249.669998  

[348 rows x 50 columns]

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 348 entries, 0 to 347
Data columns (total 50 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date           348 non-null    datetime64[ns]
 1   ADANIENT.NS    348 non-null    float64       
 2   ADANIPORTS.NS  348 non-null    float64       
 3   APOLLOHOSP.NS  348 non-null    float64       
 4   ASIANPAINT.NS  348 non-null    float64       
 5   AXISBANK.NS    348 non-null    float64       
 6   BAJAJ-AUTO.NS  348 non-null    float64       
 7   BAJAJFINSV.NS  348 non-null    float64       
 8   BAJFINANCE.NS  348 non-null    float64       
 9   BHARTIARTL.NS  348 non-null    float64       
 10  BPCL.NS        348 non-null    float64       
 11  BRITANNIA.NS   348 non-null    float64       
 12  CIPLA.NS       348 non-null    float64       
 13  COALINDIA.NS   348 non-null    float64       
 14  DIVISLAB.NS    348 non-null    float64       
 15  DRREDDY.NS     348 non-

### Convert Date column to DateTime

In [6]:
data['Date'] = pd.to_datetime(data['Date'])

### Data Cleaning & EDA

In [7]:
data.isnull().sum()

Date             0
ADANIENT.NS      0
ADANIPORTS.NS    0
APOLLOHOSP.NS    0
ASIANPAINT.NS    0
AXISBANK.NS      0
BAJAJ-AUTO.NS    0
BAJAJFINSV.NS    0
BAJFINANCE.NS    0
BHARTIARTL.NS    0
BPCL.NS          0
BRITANNIA.NS     0
CIPLA.NS         0
COALINDIA.NS     0
DIVISLAB.NS      0
DRREDDY.NS       0
EICHERMOT.NS     0
GRASIM.NS        0
HCLTECH.NS       0
HDFCBANK.NS      0
HDFCLIFE.NS      0
HEROMOTOCO.NS    0
HINDALCO.NS      0
HINDUNILVR.NS    0
ICICIBANK.NS     0
INDUSINDBK.NS    0
INFY.NS          0
ITC.NS           0
JSWSTEEL.NS      0
KOTAKBANK.NS     0
LT.NS            0
M&M.NS           0
MARUTI.NS        0
NESTLEIND.NS     0
NTPC.NS          0
ONGC.NS          0
POWERGRID.NS     0
RELIANCE.NS      0
SBILIFE.NS       0
SBIN.NS          0
SHREECEM.NS      0
SUNPHARMA.NS     0
TATAMOTORS.NS    0
TATASTEEL.NS     0
TCS.NS           0
TECHM.NS         0
TITAN.NS         0
ULTRACEMCO.NS    0
UPL.NS           0
WIPRO.NS         0
dtype: int64

### Highest growth rate for Investing

In [8]:
all_companies = data.columns[1:]

growth_on_all_stocks = data[all_companies].pct_change() * 100

average_growth_on_all_stocks = growth_on_all_stocks.mean()

average_growth_on_all_stocks.sort_values(ascending = False).head(10)

BHARTIARTL.NS    0.187065
M&M.NS           0.183317
DIVISLAB.NS      0.167704
BPCL.NS          0.136326
ADANIPORTS.NS    0.121613
ICICIBANK.NS     0.117280
BAJFINANCE.NS    0.110031
EICHERMOT.NS     0.096652
SUNPHARMA.NS     0.095606
SBIN.NS          0.093897
dtype: float64

### Highest risk companies for Investing

In [9]:
volatility_all_stocks = data[all_companies].std()

volatility_all_stocks.sort_values(ascending = False).head(10)


SHREECEM.NS      1883.955631
BAJAJ-AUTO.NS    1188.542126
DIVISLAB.NS       967.164528
ULTRACEMCO.NS     801.943227
MARUTI.NS         792.046761
HEROMOTOCO.NS     616.394425
EICHERMOT.NS      520.480751
M&M.NS            477.576008
APOLLOHOSP.NS     444.294207
BRITANNIA.NS      442.831837
dtype: float64

### Highest return on Investments

In [10]:
all_initial_stock_prices = data[all_companies].iloc[0]
all_final_stock_prices = data[all_companies].iloc[-1]

roi_on_all_stocks = ((all_final_stock_prices - all_initial_stock_prices) / all_initial_stock_prices) * 100

roi_on_all_stocks.sort_values(ascending = False).head(10)

BHARTIARTL.NS    84.249962
M&M.NS           76.046276
DIVISLAB.NS      69.886415
BPCL.NS          48.254256
ICICIBANK.NS     45.912211
BAJFINANCE.NS    39.225433
ADANIPORTS.NS    37.323799
SUNPHARMA.NS     35.109179
EICHERMOT.NS     33.534607
SBIN.NS          31.364490
dtype: float64

## Creating a Mutual Fund Plan Based on High ROI and Low Risk

To design an effective strategy for selecting companies with high returns and controlled risk, we combine Return on Investment (ROI) and volatility (standard deviation) as key evaluation metrics. The objective is to identify stocks that deliver strong long-term returns while maintaining relatively stable price movements.

The mutual fund creation process follows these steps:

**1. Define ROI and Volatility Thresholds**

Establish benchmark values for ROI and volatility to filter companies that demonstrate above-average returns with below-average risk.

**2. Rank Companies by Performance and Risk**

Rank all companies based on their ROI and volatility metrics to evaluate their growth potential and stability.

**3. Assign Investment Weights**

Allocate higher investment weights to companies with lower volatility, ensuring risk is balanced across the portfolio while preserving return potential.

We begin by defining ROI and volatility thresholds and selecting companies that satisfy the criteria of high return and low risk, forming the foundation of the mutual fund portfolio.

In [11]:
roi_threshold = roi_on_all_stocks.median()
volatility_threshold = volatility_all_stocks.median()

selected_companies = roi_on_all_stocks[
    (roi_on_all_stocks > roi_threshold) &
    (volatility_all_stocks < volatility_threshold)
]

selected_companies.sort_values(ascending = False)


BPCL.NS          48.254256
ICICIBANK.NS     45.912211
BAJFINANCE.NS    39.225433
ADANIPORTS.NS    37.323799
SBIN.NS          31.364490
POWERGRID.NS     27.946002
SBILIFE.NS       27.010009
ONGC.NS          24.561460
HDFCLIFE.NS      20.162148
CIPLA.NS         18.155687
TATASTEEL.NS     17.457401
dtype: float64

To balance investments, we use inverse volatility weighting, where companies with lower volatility receive higher weights. This helps control risk while allocating investments across selected companies.

In [12]:
selected_vol = volatility_all_stocks[selected_companies.index]

inverse_vol = 1 / selected_vol

investment_ratios = inverse_vol/inverse_vol.sum()

investment_percentages = investment_ratios * 100

(investment_percentages.sort_values(ascending=False))

TATASTEEL.NS     29.325540
ONGC.NS          14.973949
POWERGRID.NS     12.885990
BPCL.NS          12.536206
HDFCLIFE.NS       6.391885
SBIN.NS           6.292537
CIPLA.NS          4.595480
BAJFINANCE.NS     4.437315
ICICIBANK.NS      3.124709
ADANIPORTS.NS     2.850611
SBILIFE.NS        2.585779
dtype: float64

# Analyzing Our Mutual Fund Plan

After building the mutual fund for long-term investing, we evaluate its performance by comparing it with high-growth companies in the market. We begin by comparing the risk (volatility) of the mutual fund portfolio with the risk of high-performing stocks.

In [13]:
from tempfile import template
top_roi_stocks = average_growth_on_all_stocks.sort_values(ascending = False).head(5)

volatility_top_stocks = volatility_all_stocks[top_roi_stocks.index]

volatility_selected_stocks = volatility_all_stocks[selected_companies.index]

import plotly.graph_objs as go
import plotly.express as px


fig = go.Figure()

fig.add_trace(go.Bar(
    y= volatility_top_stocks.index,
    x= volatility_top_stocks,
    orientation='h',
    name='Top ROI Stocks',
    marker=dict(color='blue')
))

fig.add_trace(go.Bar(
    y= volatility_selected_stocks.index,
    x = volatility_selected_stocks,
    orientation='h',
    name='Selected Stocks',
    marker=dict(color='green')
))

fig.update_layout(
    title='Comparison of Volatility between Selected and Top Stocks',
    xaxis_title='Volatility',
    yaxis_title='Stock',
    barmode= 'overlay',
    legend=dict(title='Company Type'),
    template='plotly_white'
)

fig.show()

Now, let’s compare the ROI of both the groups as well:

In [14]:
roi_top_stocks = roi_on_all_stocks[top_roi_stocks.index]
roi_selected_stocks = roi_on_all_stocks[selected_companies.index]

fig = go.Figure()

fig.add_trace(go.Bar(
    y= roi_top_stocks.index,
    x= roi_top_stocks,
    orientation='h',
    name='Top ROI Stocks',
    marker=dict(color='blue')
))

fig.add_trace(go.Bar(
    y= roi_selected_stocks.index,
    x = roi_selected_stocks,
    orientation= 'h',
    name='Selected Stocks',
    marker=dict(color='green')
))

fig.update_layout(
    title='ROI Comparison Between Selected and Top stocks',
    xaxis_title='ROI',
    yaxis_title='Stock',
    barmode='overlay',
    legend=dict(title='Company Type'),
    template='plotly_white'
)

fig.show()

The comparison of risk (volatility) and expected ROI highlights a clear trade-off between mutual fund companies and high-growth companies. Mutual fund companies (blue) show lower volatility, making them less risky, but they also deliver moderate returns. In contrast, high-growth companies (green) exhibit higher volatility, indicating greater risk, while offering higher potential returns, as seen in stocks like Bajaj Auto and Bajaj Finserv.

For long-term investing, the objective is to balance risk and return. The companies selected in the mutual fund demonstrate stable performance with manageable risk, making them suitable for investors seeking consistent growth without large fluctuations.

# Calculating Expected Returns

Now, let’s calculate the expected returns a person will get from our mutual fund if he/she invests ₹5000 every month.

To calculate the expected value a person will accumulate over 1 year, 3 years, 5 years, and 10 years through the mutual fund plan, we can follow these steps:

1. Assume the person is investing 5000 rupees every month & increasing the investment amount by 10% every year.
2. Use the expected ROI from the mutual fund companies to simulate the growth over time.
3. Compute the compounded value of the investments for each period (1y, 3y, 5y, and 10y).
4. Visualize the accumulated value over these periods.

In [29]:
def sip_future_value(monthly_investment, monthly_return, years, step_up=0.10):
  total_value = 0
  current_monthly = monthly_investment
  for year_count in range(1, years + 1):
    for month in range(12):
      total_value = (total_value + current_monthly) * (1 + monthly_return)
    current_monthly *= (1 + step_up)
  return total_value


selected_stocks = selected_companies.head(5)
expected_annual_return = roi_on_all_stocks[selected_companies.index].mean() / 100
monthly_return = expected_annual_return / 12

print(expected_annual_return)
print(monthly_return)

monthly_investment = 5000
years_list = [1,3,5,10]

sip_results = {}

for year in years_list:
  sip_results[f"{year} Years"] = sip_future_value( monthly_investment, monthly_return, year)

print(sip_results)

0.3067026320184937
0.025558552668207807
{'1 Years': np.float64(70964.49785668765), '3 Years': np.float64(321583.4625712098), '5 Years': np.float64(821073.3669189195), '10 Years': np.float64(5054931.055935606)}


In [28]:

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=[f"{year} year{'s' if year > 1 else ''}" for year in years_list],
    y=[sip_results[f"{year} Years"] for year in years_list],
    mode='lines+markers',
    line=dict(color='blue'),
    marker=dict(size=8),
    name='Future Value'
))

fig.update_layout(
    title="Expected Value of Monthly SIP with 10% Annual Increase (₹5000 Starting)",
    xaxis_title="Investment Period",
    yaxis_title="Future Value (₹)",
    template="plotly_white",
    hovermode="x unified"
)

fig.show()


After 1 year, the accumulated value is around ₹71,000 and by 5 years, it grows to over ₹821,100. The long-term benefit is evident, with the investment growing to nearly ₹5,054,931 over 10 years, which emphasises the value of consistent investing and compounding over time for long-term investors.

# Summary

## Project Report: Mutual Fund Plan with Python
## ME_id: ACTIVITY_PROJECT_DEMO_MUTUAL_FUND_PLAN_PYTHON

### Headline
Design a data-driven mutual fund plan using Nifty 50 closing prices by combining high returns with controlled risk.


### Overview

- Objective: Build a long-term mutual fund investment strategy by identifying stocks with high ROI and low volatility using Python.

- Data Preparation: Loaded and validated Nifty 50 closing price data, ensuring there were no missing values.

- Analysis: Computed volatility, growth rate, and ROI for each company to evaluate risk and return characteristics.

- Insights: Stocks with high ROI and lower volatility are better suited for stable, long-term investments.

- Deliverables: A mutual fund portfolio with recommended investment weights and a projection of expected returns over different investment horizons.

## Milestone:
### Extended Overview

- Data Collection & Cleaning: Imported Nifty 50 closing price data from a CSV file, converted the date column to datetime format, and verified data completeness.
- Exploratory Analysis: Measured stock volatility using standard deviation, calculated growth rates using percentage change, and determined total ROI for each company.
- Visualization: Plotted stock price trends over time and compared the risk–return profiles of mutual fund companies against high-growth stocks using bar charts.
- Trend Analysis: Identified that high-growth stocks generally exhibit higher volatility, while selected mutual fund stocks show lower volatility with moderate, stable returns—making them suitable for long-term investing.


Summary: Constructed a mutual fund portfolio by selecting stocks with ROI above the median and volatility below the median. Investment weights were assigned using inverse volatility, and SIP-based return projections demonstrated the impact of compounding over 1, 3, 5, and 10 years.